Scala 2.12

Ennosigaeon opened the pull request:
**Pull Request checklist**

- [ ] The commit(s) message(s) follows the contribution [guidelines](doc/contribution-guidelines.md#commit-message-format) ?
- [ ] Tests for the changes have been added (for bug fixes / features) ?
- [ ] Docs have been added / updated (for bug fixes / features) ?

**Current behavior :** Partial Solution for  #1129

**New behavior :** Same behavior as previously but with scala-2.12.

**Other information**:
I migrated to scala-2.12. This also implies replacing spray with akka-http and updating the following libraries:
akka 2.4.20 -> 2.5.32
slick 3.1.1 -> 3.3.3
cassandra-connector 2.4.0 -> 2.4.3

To verify that that the behavior did not change, I tested the following:
- Unit tests via `sbt test`
- Manual test of routes documented in Readme.md
  - Get all/single binaries, upload binary, delete binary
  - Get all/single context, create spark context, create java context, delete context, shutdown all contexts via reboot
  - Get jobs, start async job, start sync job, start job with existing context, delete job, get job config
  - Get all data, upload data file, delete data file, delete all files via reboot
- HTTPS serving with self-signed certificate

I know that #1131 also replaces spray with akka-http. But as this PR is quite old and also changed a lot of things, I decided to start all over again and just copy single changes from that PR. I tried to keep the diff in this PR as small as possible to make merging this change easier. So there may still be some points were additional cleanup is possible.

I also shortly tested sbt-1.4.4 locally and everything seemed to work. If you would like to, I can also bump the sbt version

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/spark-jobserver/spark-jobserver/1332)
<!-- Reviewable:end -->

velvia commented:
Thank you for the PR!

> On Dec 6, 2020, at 4:22 AM, Marc <notifications@github.com> wrote:
> 
> Pull Request checklist
> 
>  The commit(s) message(s) follows the contribution guidelines <x-msg://93/doc/contribution-guidelines.md#commit-message-format> ?
>  Tests for the changes have been added (for bug fixes / features) ?
>  Docs have been added / updated (for bug fixes / features) ?
> Current behavior : Partial Solution for #1129 <https://github.com/spark-jobserver/spark-jobserver/issues/1129>
> New behavior : Same behavior as previously but with scala-2.12.
> 
> Other information:
> I migrated to scala-2.12. This also implies replacing spray with akka-http and updating the following libraries:
> akka 2.4.20 -> 2.5.32
> slick 3.1.1 -> 3.3.3
> cassandra-connector 2.4.0 -> 2.4.3
> 
> To verify that that the behavior did not change, I tested the following:
> 
> Unit tests via sbt test
> Manual test of routes documented in Readme.md
> Get all/single binaries, upload binary, delete binary
> Get all/single context, create spark context, create java context, delete context, shutdown all contexts via reboot
> Get jobs, start async job, start sync job, start job with existing context, delete job, get job config
> Get all data, upload data file, delete data file, delete all files via reboot
> HTTPS serving with self-signed certificate
> I know that #1131 <https://github.com/spark-jobserver/spark-jobserver/pull/1131> also replaces spray with akka-http. But as this PR is quite old and also changed a lot of things, I decided to start all over again and just copy single changes from that PR. I tried to keep the diff in this PR as small as possible to make merging this change easier. So there may still be some points were additional cleanup is possible.
> 
> This change is  <https://reviewable.io/reviews/spark-jobserver/spark-jobserver/1332>
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/spark-jobserver/spark-jobserver/pull/1332 <https://github.com/spark-jobserver/spark-jobserver/pull/1332>
> Commit Summary
> 
> feat: Migrate to scala-2.12
> refractor: Replace deprecated slick.driver with slick.jdbc
> refractor: Close open ActorSystems
> misc: Remove unsupported jvm opts
> File Changes
> 
> M .jvmopts <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-861f59b57142ee5ebac06ba115986da05106b4471d50848e5c21d9c58195caa1> (1)
> M Dockerfile.test <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-5faf10fde2b24acdf4f05e70be1be9d50921d685e64f781d242e056b11d79dcc> (2)
> M README.md <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5> (6)
> M akka-app/src/main/scala/spark/jobserver/common/akka/actor/Reaper.scala <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-b4f85607b1f9a667a420155aaff6d0b181444f3811884c5b75c503b0379f7b5a> (2)
> M akka-app/src/main/scala/spark/jobserver/common/akka/web/CommonRoutes.scala <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-4f7cf92d5b4ed9cc38cf4d6c990293e6e01e25a601283af25a75e063f96ae166> (7)
> M akka-app/src/main/scala/spark/jobserver/common/akka/web/WebService.scala <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-2f149c69720d620a422f2450eb894b3196e4d1b3bc26c878f00c4b0c90e60cc1> (29)
> M akka-app/src/test/scala/spark/jobserver/common/akka/actor/ReaperSpec.scala <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-fbd8e2ba843d7934e83745ec91c672003594313efc56b766cd38123aafff0538> (2)
> M akka-app/src/test/scala/spark/jobserver/common/akka/web/CommonRoutesSpec.scala <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-13f08e2e0e2096ba49c3ed6dd8d2a85ae2dbd41bb80a451150640b7f57abc760> (9)
> M bin/setenv.sh <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-10d0d624ae86872fd945583a677d59cbd18dede48567b89a9f4f2326b1432715> (1)
> M build.sbt <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-5634c415cd8c8504fdb973a3ed092300b43c4b8fc1e184f7249eb29a55511f91> (3)
> M doc/EMR.md <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-2ec6523c70454dc76b3f38a2040be9ce94000cafd78e1e96ce72bdb0c7588111> (2)
> M doc/dao-setup.md <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-3a95368d00a3874a579c9edf6f4940352e1b0279600c17f015ff2fc7a723effa> (8)
> M doc/troubleshooting.md <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-6f21344470815bff488d18edd851993d6b4c0cff71bfa0c1dfbd356e98a44a73> (2)
> M job-server/config/ec2.conf.template <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-dbf843603428f2ee00bc61b5e1106e2c592d55a6402006952d2f5b498e614dd9> (2)
> M job-server/config/local.conf.template <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-121b8a4dc8a25a0dcb007a107e947cbf4c3148af8dc2c1cba0f56056f699ba28> (2)
> M job-server/src/main/resources/application.conf <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-6cefda4cb9c8b4b00cc1f35d20e088c17cb31fbd2547fc67685a9aade065fe35> (6)
> M job-server/src/main/scala/spark/jobserver/AkkaClusterSupervisorActor.scala <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-b0d7e0693b3a82f2b32ced056779e214f7d26d108dd370af5afbcc94a85cd2fc> (6)
> M job-server/src/main/scala/spark/jobserver/ChunkEncodedStreamingSupport.scala <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-fe2cb45669f66d1c8f5bb496f839fff9c96af1b21f26dc87832acf38a8a21b24> (36)
> D job-server/src/main/scala/spark/jobserver/ChunkEncodingActor.scala <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-e200b306c8400977225ea61f7f2c9c8ba1f4a540a1c67c219395e25a3462417e> (48)
> D job-server/src/main/scala/spark/jobserver/CorsSupport.scala <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-85b81b10ad72a42b9542f944dbfe660359e8b43a34ad0985e260ccaf3aaadb4a> (40)
> M job-server/src/main/scala/spark/jobserver/JobManagerActor.scala <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-dc65fa0360d8e06a217b88bab7a4761e233330765de7404a14fcdf32ea04fc2e> (2)
> M job-server/src/main/scala/spark/jobserver/JobServer.scala <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-2fe7512db81b5c0892d2e00ffd241e20b1d3eff3bcf1be5ad4904b4e1de99ad4> (2)
> M job-server/src/main/scala/spark/jobserver/JobServerNamedObjects.scala <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-1cb579d10c38a901f3f184e7d96cd5c71cebd4799949e4f7eaa01eaa969db8be> (38)
> M job-server/src/main/scala/spark/jobserver/WebApi.scala <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-32dc7248a2ea910c7efdb7b91aa410508b0071b09e745d4cc2970020d83f8c4b> (265)
> M job-server/src/main/scala/spark/jobserver/auth/SJSAuthenticator.scala <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-758be88fbf1568c9d87677f09951ce7784e0c80b60f521f7ee2467f843de9313> (87)
> M job-server/src/main/scala/spark/jobserver/io/InfoDefinitions.scala <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-7c9c9d4755e41e587df786f22e62128f93767b1f6b0d3e85b31dc0fab7f86192> (26)
> M job-server/src/main/scala/spark/jobserver/routes/DataRoutes.scala <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-a1096f5fad65b2e0bce33e25ccd931c140a6e68344f8655296059a0ff7536633> (45)
> M job-server/src/main/scala/spark/jobserver/slick/unmanaged/UnmanagedDatabase.scala <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-575bb5ff776a94efc5a5191cb7a9339ae49cd4fd4274c9d09718ca8de41e3c42> (1)
> M job-server/src/main/scala/spark/jobserver/util/ForcefulKill.scala <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-fdbaf292dfb6d72078447a77f85f2d39ca271cfdb6cb86777178ca0287804218> (62)
> M job-server/src/main/scala/spark/jobserver/util/HealthCheck.scala <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-a1985d7265d184ac712978ae7501da042a151043a826221101f438a361584e6d> (13)
> M job-server/src/main/scala/spark/jobserver/util/MeteredHttpService.scala <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-ee9895bbae5460930457ca4577d066225e821248b3e8522096b2979b9412a67e> (32)
> M job-server/src/main/scala/spark/jobserver/util/SqlDBUtils.scala <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-0360e4134c822648f8b5de9794e0af962e426bf718a0d4213aa7de55a5a4bd8f> (2)
> M job-server/src/test/resources/local.test.binarysqldao.conf <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-c118726319521500e0ee0d8f124d6046bf763adda030c0f162dbbfaa6788dfeb> (2)
> M job-server/src/test/resources/local.test.metadatasqldao.conf <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-eed0ff7254a0060e2864bbc7fcb3aec462f0237d2ba3a12c77179713e468b356> (2)
> M job-server/src/test/scala/akka/cluster/TestMember.scala <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-91d0232eed5a78f584280e914e915e5e15b132149e39f5b65b3d992469e88050> (3)
> M job-server/src/test/scala/spark/jobserver/AkkaClusterSupervisorActorSpec.scala <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-d3d3f4331c4ff7e71e2632524df318a7e9595ef8c2df994ce09e69f20e9ddb41> (2)
> M job-server/src/test/scala/spark/jobserver/HealthzApiSpec.scala <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-305e7bb47f9cd410de48fdfb6aaa4a475f7541172061fe060c08e08c424f9c75> (19)
> M job-server/src/test/scala/spark/jobserver/JobWithNamedRddsSpec.scala <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-96f9df64f9bdd2471d8a36a2a0239d1b169f98b9ecd48701a9bb3886dc5b454e> (21)
> M job-server/src/test/scala/spark/jobserver/WebApiMainRoutesSpec.scala <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-d8cf4fbc6b6232357392dbe10ae15f201d8c276f19d5f0a9c2cb99990da14774> (41)
> M job-server/src/test/scala/spark/jobserver/WebApiSpec.scala <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-22dde887eab01bdab7e2e991b3f459dada2b02b2d1c6c611fdc4583ef0b804ec> (10)
> M job-server/src/test/scala/spark/jobserver/auth/SJSAuthenticatorSpec.scala <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-472548094fbbcc42c185a5bfca245c1bb9c5bd33cb0a3e6c054d7d7b129d90e6> (30)
> M job-server/src/test/scala/spark/jobserver/auth/WebApiWithAuthenticationSpec.scala <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-d22a76f4ef03993cc592f1b76c7f2aa4843dc694ed22e1ea40bfb6df7458decd> (55)
> M job-server/src/test/scala/spark/jobserver/io/zookeeper/AutoPurgeActorSpec.scala <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-73266f86a286db79601a1fa4b67fa0c498855454f5188f2151c66a8febb21af0> (10)
> M job-server/src/test/scala/spark/jobserver/routes/DataRoutesSpec.scala <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-c89d421e78d50da8c0a61a3f68a1dfa6c10ac5fb614aee2a0f914b8e170fc300> (7)
> M job-server/src/test/scala/spark/jobserver/util/ForcefulKillSpec.scala <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-fd3950b6c53afd3f992cb076e64f3d09c653b9f1bf704f95d9ae7f2b39e18d42> (31)
> M job-server/src/test/scala/spark/jobserver/util/JsonProtocolsSpec.scala <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-fa34e38fd0e0dfca2b2b1f7029afdd713fcb5830a11f9937a86943fd72fd2098> (24)
> M job-server/src/test/scala/spark/jobserver/util/SSLContextFactorySpec.scala <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-083c1778ee2bb41e5226faee7c29ece402883b6e6d4acb3a4e289700bfc22662> (28)
> M project/Dependencies.scala <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-ad2642dc77b3679e4ad2c7d3b2f59a2dcf48c1bf5b93a558a4fa44828315a34c> (20)
> M project/Versions.scala <https://github.com/spark-jobserver/spark-jobserver/pull/1332/files#diff-34c3448479f3e781c441f9b9dc28e3aa12f50b80ceb426d6dabf54b6751b84d4> (9)
> Patch Links:
> 
> https://github.com/spark-jobserver/spark-jobserver/pull/1332.patch <https://github.com/spark-jobserver/spark-jobserver/pull/1332.patch>
> https://github.com/spark-jobserver/spark-jobserver/pull/1332.diff <https://github.com/spark-jobserver/spark-jobserver/pull/1332.diff>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub <https://github.com/spark-jobserver/spark-jobserver/pull/1332>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAIDPWZXC3I26VSON3Q5ZL3STNZPZANCNFSM4UPI652Q>.
>

Ennosigaeon commented:
I will take another look at the single failing test case later today. Locally, all unit tests passed...

Ennosigaeon commented:
Ok, there are actually two failing unit tests: `JobManagerActorSpec` and `AkkaClusterSupervisorActorSpec`. I am not able to reproduce them locally, neither via `sbt test` nor `sbt job-sever/testOnly *[...]`. If I update to sbt-1.4.4 even more tests fail in travis that do not fail locally. I am not sure how to procede with these failures... Maybe somebody else can run `sbt test` locally.

valan4ik commented:
Hi @Ennosigaeon ,

Thanks a lot for the PR!
I started to look into it, but will probably need several days to finish the review. Need to read upgrade guides and test changes.

I also restarted the tests and they are now green. `AkkaClusterSupervisorActorSpec` can be a bit flaky.

valan4ik reviewed (changes requested):
We should enable Scala 2.11 support at least for one more major release of Jobserver. Spark 2.4 supports both 2.11 and 2.12 (and by default compiled with 2.11).

Ennosigaeon commented on job-server/src/test/scala/akka/cluster/TestMember.scala:
```diff
@@ -10,5 +10,6 @@ object TestMember {
   def apply(address: Address, status: MemberStatus, roles: Set[String]): Member =
-    new Member(UniqueAddress(address, 0), Int.MaxValue, status, roles)
+    new Member(UniqueAddress(address, 0), Int.MaxValue, status,
+      roles ++ Set("dc-default"))
```
Not completely sure about this one. As far as I understand, akka-2.5 introduced support for multiple data centers. Members have to provide a `dc-*` role (see [member implementation](https://github.com/akka/akka/blob/01e4b4de27df51b25077043df601bc1ffb57315d/akka-cluster/src/main/scala/akka/cluster/Member.scala#L34)). According to this [akka issue](https://github.com/akka/akka/issues/25496) the dc-role should be set automatically (which it isn't in our test case) to `dc-default` with nothing is provided. As this only affects test cases, I did not dig further into it.

bsikander replied:
Interesting, as you mentioned, this is fixed in the version that we are using but somehow it breaks for tests.

suggestion:
For a wider audience, a comment on top of this line would make sense to tell them what this is.

Ennosigaeon commented on job-server/src/main/scala/spark/jobserver/slick/unmanaged/UnmanagedDatabase.scala:
```diff
@@ -11,6 +11,7 @@ import slick.util.AsyncExecutor
 class UnmanagedDataSource(conn: Connection) extends JdbcDataSource {
   def createConnection(): Connection = conn
   def close(): Unit = ()
+  val maxConnections: Option[Int] = None
```
I had to update slick to get scala-2.12 support and the updates `JdbcDataSource` requires this attribute. `None` effectively does not limit the maximum number of connections.

bsikander replied:
ok but is this overriding something? If yes, then maybe we should explicitly write `override` keyword.

Ennosigaeon replied:
No override, the [JdbcDataSource](https://github.com/slick/slick/blob/v3.3.2/slick/src/main/scala/slick/jdbc/JdbcDataSource.scala) requires it to be implemented. In the previous slick version that field did not exist

Ennosigaeon commented on job-server/src/main/scala/spark/jobserver/WebApi.scala:
```diff
@@ -335,39 +328,42 @@ class WebApi(system: ActorSystem,
                       val future = binaryManager ? StoreBinary(appName, binaryType, binBytes)
 
-                      future.map {
+                      future.flatMap {
```
The akka-http [Request.Complete](https://doc.akka.io/japi/akka-http/current/akka/http/scaladsl/server/RequestContext.html#complete-akka.http.scaladsl.marshalling.ToResponseMarshallable-) returns a future instead of synchronously producing a result like [spray did](https://github.com/spray/spray/blob/da11335d03898b36040c1ad358084d371b1795d0/spray-routing/src/main/scala/spray/routing/RequestContext.scala#L224). To keep the remaining code unchanged (I had some trouble to prevent _accidental_ indentation changes to prevent the diff from exploding), I resorted to use flatMaps everywhere.

Ennosigaeon commented on job-server/src/main/scala/spark/jobserver/CorsSupport.scala:
```diff
@@ -1,40 +0,0 @@
-package spark.jobserver
-
-import spray.http._
-import spray.http.HttpHeaders._
-import spray.http.HttpMethods._
-import spray.routing._
-
-// origin of code:
-// first example: https://gist.github.com/joseraya/176821d856b43b1cfe19
-// improvement: https://gist.github.com/waymost/4b5598523c2c7361abea
-// see also https://developer.mozilla.org/en-US/docs/Web/HTTP/Access_control_CORS
-trait CORSSupport {
-  this: HttpService =>
-
-  private val allowOriginHeader = `Access-Control-Allow-Origin`(AllOrigins)
-  private val optionsCorsHeaders = List(
-    `Access-Control-Allow-Headers`("Origin, X-Requested-With, Content-Type, Accept, Accept-Encoding, " +
-      "Accept-Language, Host, Referer, User-Agent"),
-    `Access-Control-Max-Age`(1728000)
-  )
-
-  def cors[T]: Directive0 = mapRequestContext { ctx =>
-    ctx.withRouteResponseHandling {
-      // OPTION request for a resource that responds to other methods
-      case Rejected(x) if (ctx.request.method.equals(HttpMethods.OPTIONS)
-        && x.exists(_.isInstanceOf[MethodRejection])) => {
-        val allowedMethods: List[HttpMethod] =
-          x.collect { case rejection: MethodRejection => rejection.supported }
-        ctx.complete {
-          HttpResponse().withHeaders(
-            `Access-Control-Allow-Methods`(HttpMethods.OPTIONS, allowedMethods: _*) ::
-              allowOriginHeader :: optionsCorsHeaders
-          )
-        }
-      }
-    }.withHttpResponseHeadersMapped { headers =>
-      allowOriginHeader :: headers
-    }
-  }
-}
```
I have added `"ch.megard" %% "akka-http-cors" % "0.4.2"` to handle CORS headers (WebApi@219). Headers are still set:

```
curl -H "Origin: http://example.com" -v localhost:8090/jobs
*   Trying 127.0.0.1:8090...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 8090 (#0)
> GET /jobs HTTP/1.1
> Host: localhost:8090
> User-Agent: curl/7.68.0
> Accept: */*
> Origin: http://example.com
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Access-Control-Allow-Origin: http://example.com
< Access-Control-Allow-Credentials: true
< Server: akka-http/10.1.13
< Date: Wed, 16 Dec 2020 08:23:21 GMT
< Content-Type: application/json
< Content-Length: 2
< 
* Connection #0 to host localhost left intact
[]%  
```

valan4ik replied:
Oh, I compared only the output headers. Previously we were always returning `Access-Control-Allow-Origin: *`, and now it is returned only if  `Origin` header is provided (also content is also different). 
But as the module is there, it is now easily configurable. I played with a settings a bit, seems good.

Ennosigaeon commented on build.sbt:
```diff
@@ -262,7 +261,7 @@ lazy val commonSettings = Defaults.coreDefaultSettings ++ dirSettings ++ Seq(
   crossPaths   := true,
   releaseIgnoreUntrackedFiles := true,
-  scalaVersion := sys.env.getOrElse("SCALA_VERSION", "2.11.8"),
+  scalaVersion := sys.env.getOrElse("SCALA_VERSION", "2.12.12"),
```
I have added cross compile for Scala 2.11 and 2.12. Is the option to override the Scala version via env variables still necessary? If so, I can add support again.

valan4ik replied:
Yes, we should also have an option to override the Scala version (`crossScalaVersions` is for releases only). From my point of view, this line should be kept.

Ennosigaeon replied:
Ok, I will add it again.

Ennosigaeon replied:
Sorry, I forgot to actually push the change...

Ennosigaeon commented:
Hey, I tried to resolve most of your comments. For better readability, they are split into multiple commits, but I will squash them when you approve the changes.

Ennosigaeon commented on job-server/src/main/scala/spark/jobserver/AkkaClusterSupervisorActor.scala:
```diff
@@ -449,13 +449,13 @@ class AkkaClusterSupervisorActor(daoActor: ActorRef, dataManagerActor: ActorRef,
       case Failure(e: Exception) =>
         logger.info("Failed to send initialize message to context " + ref, e)
-        cluster.down(ref.path.address)
+        context.stop(ref)
```
I am not sure what the exact reasons are, but somehow the behavior between akka-2.4 and akka-2.5 changed. `ref.path.address` contains not the address of the actual actor. In the actor supervisor spec,  the address was `akka://test` shutting the complete cluster down instead of only a single actor.

valan4ik commented:
@Ennosigaeon ,

I now tested your changes with Scala 2.11 and found one more problem when running HA tests from `job-server-integration-tests` subproject.

Please switch off WeaklyUp members (they are enabled by default in Akka 2.5), having WeaklyUp members setting on would definitely make Jobserver unstable.
Add to Akka section in `application.conf`:

```
  cluster {
    allow-weakly-up-members = "off"
  }
```


I also found problems with new `coordinated-shutdown` feature, I had to add
```
  coordinated-shutdown {
    exit-jvm = "on"
    run-by-jvm-shutdown-hook = "off"
  }
```
to make tests work. I will investigate this problem later, don't think that it should be a blocker for your PR.


Except for 2 problems above, all tests seem to be green and HA and SSL setup seems to work.
Thank you very much for the PR and for implementing all requested changes so fast.

Ennosigaeon commented:
@valan4ik: Just let me know when you are happy with all changes. I will squash all commits than to conform with the commit name requirements.

valan4ik commented:
LGTM from my side.
@bsikander are all your questions also solved?

bsikander commented:
> @Ennosigaeon ,
> 
> I now tested your changes with Scala 2.11 and found one more problem when running HA tests from `job-server-integration-tests` subproject.
> 
> Please switch off WeaklyUp members (they are enabled by default in Akka 2.5), having WeaklyUp members setting on would definitely make Jobserver unstable.
> Add to Akka section in `application.conf`:
> 
> ```
>   cluster {
>     allow-weakly-up-members = "off"
>   }
> ```
> 
> I also found problems with new `coordinated-shutdown` feature, I had to add
> 
> ```
>   coordinated-shutdown {
>     exit-jvm = "on"
>     run-by-jvm-shutdown-hook = "off"
>   }
> ```
> 
> to make tests work. I will investigate this problem later, don't think that it should be a blocker for your PR.
> 
> Except for 2 problems above, all tests seem to be green and HA and SSL setup seems to work.
> Thank you very much for the PR and for implementing all requested changes so fast.

Thank you for pointing this out. The first point is exactly what I was trying to find and never remembered. +1

bsikander commented on project/Dependencies.scala:
```diff
@@ -21,11 +21,14 @@ object Dependencies {
+    "com.typesafe.akka" %% "akka-http-core" % akkaHttp,
+    "com.typesafe.akka" %% "akka-http-caching" % akkaHttp,
+    "com.typesafe.akka" %% "akka-http-spray-json" % akkaHttp,
+    "ch.megard" %% "akka-http-cors" % "0.4.2",
```
Could you put the version in the `Versions` file?

bsikander commented:
Overall looks good. Once the above comments are addressed, we can move forward to the next step.

Ennosigaeon commented:
@bsikander: I have added a comment in TestMember.scala and moved akka-http-cors and slf4j version number to Versionen. Additionally, I have squased all review commits into the scala-2.12 migration commit.

bsikander commented:
One more thing, this can be done after this change also, so you can decide.

Since with this change we support 2.12, we should add this in README. Readme already points to 2.10,2.11. You can find this at the top of readme or search "Supports Scala 2.10 and 2.11".

Ennosigaeon commented:
Ok, I have updated the Readme with Scala 2.11 and Scala 2.12 being supported. I have also adapted the `config/ec2.sh` and `config/local.sh.template` files to use Scala 2.12.12. Finally, I rebased on master to include the latest commits.

valan4ik commented:
@Ennosigaeon ,
sorry, I think my comment on `coordinated-shutdown` fixed the HA integration tests, but now unit tests are broken :( I didn't rerun them..
Could you please try to remove `exit-jvm = "on"` line?

Ennosigaeon commented:
@valan4ik: I removed `exit-jvm`, rerun unit tests with `+ test` and everything looks fine. I am not sure how to run HA tests locally...
